fix: ensure key exchange functions properly when a sync service is not being used

gkc opened the pull request:
**- What I did**
fix: resolves #1041

**- How I did it**
- Implemented new logic in abstract_atkey_encryption.dart as specified in #1041
- Tweaked mockLocalSecondary and mockRemoteSecondary setup in unit tests in encryption_service_test.dart
- Tweaked mockLocalSecondary and mockRemoteSecondary setup in some other unit tests
- No other actual changes (just fixed some unnecessary import statements from a few files)
- New test cases which verify that the key exchange is working as intended in all scenarios

**- How to verify it**
Tests should pass

**- Description for the changelog**
fix: ensure key exchange functions properly when no sync service is being used

gkc committed:
fix: resolves #1041
- Implemented new logic in abstract_atkey_encryption.dart as specified in #1041
- Tweaked mockLocalSecondary and mockRemoteSecondary setup in encryption_service_test.dart
- No other actual changes (just fixed some unnecessary import statements from a few files)

gkc committed:
style: ran dart format

gkc committed:
test: tweaked mock setup in full_stack_test.dart

gkc committed:
fix: ensured 'data:' is stripped from responses from lookups of encrypted symmetric keys

gkc committed:
test: put log level back to 'info' in bypasscache_test.dart as logs at 'finer' level were huge, and impossible to look through when tests were failing.

gkc committed:
style: dart format

gkc commented on packages/at_client/lib/src/encryption_service/abstract_atkey_encryption.dart:
```diff
@@ -29,104 +27,76 @@ abstract class AbstractAtKeyEncryption implements AtKeyEncryption {
         'AbstractAtKeyEncryption (${_atClient.getCurrentAtSign()})');
   }
 
-  @visibleForTesting
```
This was being used to try to ensure something had been sync'd. Removing as now obsolete.

gkc commented on packages/at_client/lib/src/encryption_service/abstract_atkey_encryption.dart:
```diff
@@ -29,104 +27,76 @@ abstract class AbstractAtKeyEncryption implements AtKeyEncryption {
-    if (!(await isEncryptedSharedKeyInSync(atKey))) {
-      await updateEncryptedSharedKeyToSecondary(atKey, encryptedSharedKey,
-          secondary: _atClient.getRemoteSecondary());
+      _sharedKey = await createMyCopyOfSharedSymmetricKey(atKey);
```
Created new method `createMyCopyOfSharedSymmetricKey` for readability of this (encrypt) method and clarity around the new logic

gkc commented on packages/at_client/lib/src/encryption_service/abstract_atkey_encryption.dart:
```diff
@@ -29,104 +27,76 @@ abstract class AbstractAtKeyEncryption implements AtKeyEncryption {
   /// Throws [KeyNotFoundException] if the encryptionPrivateKey is not found.
-  Future<String> getSharedKey(AtKey atKey) async {
+  ///
+  Future<String> getMyCopyOfSharedSymmetricKey(AtKey atKey) async {
```
Renamed for clarity

gkc commented on packages/at_client/lib/src/encryption_service/abstract_atkey_encryption.dart:
```diff
@@ -29,104 +27,76 @@ abstract class AbstractAtKeyEncryption implements AtKeyEncryption {
-  /// If found, returns the decrypted sharedKey.
-  ///
-  /// If shared_key is not found in local secondary and remote secondary, returns an empty string
-  ///
```
New method comment below describes the desired new logic.

gkc commented on packages/at_client/lib/src/encryption_service/abstract_atkey_encryption.dart:
```diff
@@ -159,6 +133,100 @@ abstract class AbstractAtKeyEncryption implements AtKeyEncryption {
     }
   }
 
```
New method comment below describes the desired new logic for if / when a new shared symmetric key needs to be created

gkc commented on packages/at_client/lib/src/encryption_service/abstract_atkey_encryption.dart:
```diff
@@ -201,59 +269,62 @@ abstract class AbstractAtKeyEncryption implements AtKeyEncryption {
   }
 
```
renamed _storeSharedKey and removed `shouldSync` parameter as all of this stuff is no longer dependent on sync in any way

gkc commented on packages/at_client/lib/src/encryption_service/abstract_atkey_encryption.dart:
```diff
@@ -266,6 +337,6 @@ abstract class AbstractAtKeyEncryption implements AtKeyEncryption {
       ..sharedBy = atKey.sharedBy
       ..ttr = 3888000
       ..value = encryptedSharedKeyValue;
```
`sync: false` now, because we are always saving to atServer first anyway

gkc commented on packages/at_client/lib/src/encryption_service/abstract_atkey_encryption.dart:
```diff
@@ -201,59 +269,62 @@ abstract class AbstractAtKeyEncryption implements AtKeyEncryption {
+    }
+    return myCopy;
   }
 
```
The next bit of diff here is kind of rubbish - I removed an obsolete method called isEncryptedSharedKeyInSync. Unrelated, I added a new method called _getTheirEncryptedCopyOfSharedSymmetricKey

gkc commented:
Some new unit tests to be added but is probably good enough now to start testing from other places via dependency override to this branch

gkc committed:
feat: Implemented desired logic for 'their' copy of the shared symmetric key as laid out in #1041

gkc committed:
Merge remote-tracking branch 'origin/trunk' into gkc-symmetric-key-exchange

gkc committed:
style: omg dart format

gkc committed:
Merge remote-tracking branch 'origin/trunk' into gkc-symmetric-key-exchange

gkc committed:
docs: improve clarity of addition to the changelog

gkc committed:
Merge branch 'trunk' into gkc-symmetric-key-exchange

cconstab reviewed (dismissed):
Everything as we discussed thank you...

The only possible improvement could be after getting the local AES key and if that fails to decrypt then to circle back and pull from secondary and retry if the keys are different.

I have seen this went a sync has happened in the dim past and the atSIgn has been reset for example and the local AES key is wrong. The way to get out of that loop currently is pretty harsh, delete the local storage.

This I think is now pretty unlikely/is not a problem now, but I thought I should mention it..

Colin

gkc committed:
Merge remote-tracking branch 'origin/trunk' into gkc-symmetric-key-exchange

gkc committed:
fix: Added additional defensive code around 'their' copy of shared symmetric key. We now delete 'their' copy (1) from local storage if we don't find 'our' copy in local storage, and (2) from the remote atServer when we are creating and sharing a new symmetric key
test: adjusted mock setup in encryption_service_test.dart to handle the new 'delete' requests mentioned in the previous point
test: Added new test cases to verify all the key exchange behaviour with regard to local and remote storage
build: Updated package version number and CHANGELOG

gkc commented:
Hi @sitaram-kalluri and @cconstab ... I've finally got the new unit tests to how I wanted them. Please review when you get a chance - thank you!

gkc committed:
Merge remote-tracking branch 'origin/trunk' into gkc-symmetric-key-exchange

# Conflicts:
#	tests/at_end2end_test/test/bypasscache_test.dart

purnimavenkatasubbu commented:
I did a quick sanity of the branch with apps. Attaching the logs.
[addANewContactLogSymmetricKey.txt](https://github.com/atsign-foundation/at_client_sdk/files/11705447/addANewContactLogSymmetricKey.txt)
[atMospherePro_symmetricKey_Sender_logs.txt](https://github.com/atsign-foundation/at_client_sdk/files/11705450/atMospherePro_symmetricKey_Sender_logs.txt)
[receiver_symmetric_key_logs.txt](https://github.com/atsign-foundation/at_client_sdk/files/11705453/receiver_symmetric_key_logs.txt)
[wavi_logs_symmetric_key_exchange.txt](https://github.com/atsign-foundation/at_client_sdk/files/11705455/wavi_logs_symmetric_key_exchange.txt)

https://github.com/atsign-foundation/at_client_sdk/assets/43639593/f88484e0-b7d3-43b9-90e4-742328da52d4

NOTE: Test logs of the branch before it got merged into trunk.

gkc commented:
Thank you @purnimavenkatasubbu !!! Any issues encountered?

purnimavenkatasubbu commented:
> Thank you @purnimavenkatasubbu !!! Any issues encountered?

No, Gary.

gkc commented:
Great! Thank you @purnimavenkatasubbu
